### PR TITLE
Java: Test generator improvements

### PR DESCRIPTION
**Pull request by joefarebrother**
This PR allows the test generator to accept a yml file or a diretory of yml files containing MaD specifications as input. 
Also, the generated "return type" of a constuctor is now considered to be the constructed type.

### CODE REVIEW

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py [hunk 1]
```diff
@@ -75,38 +79,86 @@
...
+        print("Failed to open %s: %s\n" % (file, e))
+        sys.exit(1)
+
+
+def readYmlDir(dirname):
+    specs = []
+    for f in os.listdir(dirname):
+        if f.endswith('.yml'):
+            specs += readYml(f"{dirname}/{f}")
+    return specs
```
Comment (atorralba) [inline java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:130]: May I suggest an implementation a bit more pythonic?

```suggestion
    return [readYml(f"{dirname}/{f}") for f in os.listdir(dirname) if f.endswith('.yml')]
```

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py [hunk 2]
```diff
@@ -75,38 +79,86 @@
           (sys.argv[2], e), file=sys.stderr)
     sys.exit(1)
 
-commentRegex = re.compile("^\s*(//|#)")
+commentRegex = re.compile(r"^\s*(//|#)")
```
Comment (atorralba) [inline java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:82]: Good catch!

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py [hunk 3]
```diff
@@ -75,38 +79,86 @@
...
+    return specs
+
+
+specsFile = sys.argv[1]
+if os.path.isdir(specsFile):
+    specs = readYmlDir(specsFile)
+elif specsFile.endswith(".yml") or specsFile.endswith(".yaml"):
+    specs = readYml(specsFile)
+elif specsFile.endswith(".csv"):
+    spcs = readCsv(specsFile)
```
Comment (atorralba) [inline java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:139]:
```suggestion
    specs = readCsv(specsFile)
```

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py [hunk 4]
```diff
@@ -221,7 +273,7 @@ def copyfile(fromName, toFileHandle):
     # Make a test extension file
     with open(resultYml, "w") as f:
         models = "\n".join('      - [%s]' %
-                          modelSpecRow[0].strip() for modelSpecRow in supportModelRows)
+                           modelSpecRow[0].strip() for modelSpecRow in supportModelRows)
```
Comment (atorralba) [inline java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:276]: Is this change intended?

**Review by atorralba — COMMENTED** [review]
Thanks for adding this! It'll be super useful for the rest of the modeling work we're doing.

I added some minor comments, but otherwise this LGTM.

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py (hunk 1 above)
Comment (joefarebrother) [reply java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:130]: That isn't equivalent is it? As that would result in a list of lists, rather than a single list of rows

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py (hunk 4 above)
Comment (joefarebrother) [reply java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:276]: This change was made by my autoformatter

File: java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py (hunk 1 above)
Comment (atorralba) [reply java/ql/src/utils/flowtestcasegenerator/GenerateFlowTestCase.py:130]: D'oh, true, sorry. That would then probably need to be something like:

```suggestion
    return [row for row in readYml(f"{dirname}/{f}") for f in os.listdir(dirname) if f.endswith('.yml')]
```

But the readability starts getting bad at that point, so feel free to leave it as it is.

**Review by atorralba — APPROVED** [review]
